### ISSUE TITLE
1.2.4 breaks aliased ID's in Unions

### ISSUE THREAD

**Issue by tlester** [neo4j/graphql#388]
Error:  

```
Cannot return null for non-nullable field Post.id.
    at new ApolloError (index.ts:49)
```

Union:
```
 query GetContent($userID: ID, $limit: Int, $skip: Int) {
        getContent(userID: $userID, limit: $limit, skip: $skip) {
            ... on Post {
                postID: id
                createdDate
                modifiedDate
                __typename
            }
            ... on Image {
                imageID: id
                createdDate
                modifiedDate
                __typename
            }
        }
        getContentCount(userID: $userID)
    }
```

id is non-nullable.... and I suspect that the aliasing isn't respecting that.

**Comment by darrellwarde** [comment]
@tlester, can we please have the type definitions? Especially the definition for `getContent`.

**Comment by tlester** [comment]
```
 getContentCount(userID: ID): Int
            @cypher(
                statement: """
                match (myUser:User {id: $userID})
                optional match (myUser)<-[:HAS_USER]-(:Profile)-[:HAS_FRIEND]->(x)<-[:HAS_FRIEND]-(:Profile)-[:HAS_USER]->(myFriends:User)
                    call {
                       with myUser, myFriends
                       match (myFriends)-[:HAS_POST]->(post:Post)
                       return post
                       union
                       with myUser, myFriends
                       match (myFriends)-[:HAS_COMMENT]->(:Comment)<-[:HAS_COMMENT]-(post:Post)
                       return post
                       union
                       with myUser, myFriends
                       match (myFriends)-[:HAS_REACTION]->(:Reaction)<-[:HAS_REACTION]-(post:Post)
                       return post
                       union
                       with myUser, myFriends
                       match (myUser)<-[:HAS_USER]-(myProfile:Profile)
                       match (myProfile)-[:HAS_USER]->(myUsers:User)
                       match (myUsers)-[:HAS_POST]->(post:Post)
                       return post
                       union
                       with myUser, myFriends
                       match (myUser)<-[:HAS_USER]-(myProfile:Profile)
                       match (myProfile)-[:HAS_USER]->(myUsers:User)
                       match (myUsers)-[:HAS_COMMENT]->(:Comment)<-[:HAS_COMMENT]-(post:Post)
                       return post
                       union
                       with myUser, myFriends
                       match (myUser)<-[:HAS_USER]-(myProfile:Profile)
                       match (myProfile)-[:HAS_USER]->(myUsers:User)
                       match (myUsers)-[:HAS_REACTION]->(:Reaction)<-[:HAS_REACTION]-(post:Post)
                       return post
                   }
                   return count(distinct post)
                """
            )
```

```
  union Content = Post | Image
```

```
    type Post @auth(rules: [{ operations: [CREATE, DELETE, UPDATE, CONNECT, DISCONNECT], isAuthenticated: true }]) {
        id: ID! @id
        hasImages: [Image] @relationship(type: "HAS_IMAGE", direction: OUT)
        body: String
        createdDate: DateTime! @timestamp(operations: [CREATE])
        modifiedDate: DateTime!
        author: User @relationship(type: "HAS_POST", direction: IN)
        hasReactions: [Reaction] @relationship(type: "HAS_REACTION", direction: OUT)
        hasComments: [Comment] @relationship(type: "HAS_COMMENT", direction: OUT)
        hasNotifications: [Notification] @relationship(type: "HAS_NOTIFICATION", direction: OUT)
        hasSharedPost: Post @relationship(type: "SHARES", direction: OUT)
        isSharedIn: [Post] @relationship(type: "SHARES", direction: IN)
    }
```

```
    type Image @auth(rules: [{ operations: [CREATE, DELETE, UPDATE, CONNECT, DISCONNECT], isAuthenticated: true }]) {
        id: String!
        body: String
        isPublic: Boolean
        signedURLs: [signedURLs]
        createdDate: DateTime! @timestamp(operations: [CREATE])
        modifiedDate: DateTime!
        inAlbum: [Album] @relationship(type: "HAS_IMAGE", direction: IN)
        inBlog: [Blog] @relationship(type: "HAS_IMAGE", direction: IN)
        author: User @relationship(type: "HAS_IMAGE", direction: IN)
        isProfilePic: Profile @relationship(type: "HAS_PROFILE_PIC", direction: IN)
        isCoverPhoto: Profile @relationship(type: "HAS_COVER_PHOTO", direction: IN)
        hasReactions: [Reaction] @relationship(type: "HAS_REACTION", direction: OUT)
        hasComments: [Comment] @relationship(type: "HAS_COMMENT", direction: OUT)
        hasNotifications: [Notification] @relationship(type: "HAS_NOTIFICATION", direction: OUT)
    }
```

**Comment by tlester** [comment]
Note that this works great in 1.2.3 and earlier.

**Comment by darrellwarde** [comment]
Luckily I've been able to reproduce from old information we have scattered around in our tests.

**Comment by tlester** [comment]
Shit...sorry.   getContent is exactly the same, but returns the nodes instead of counting them.

**Comment by cramatt** [comment]
This also breaks in simple, non-union cases in v1.2.4. For example:

```
{
  designs:teamDesigns {
    name
  	x:battery_amps
  }
}
```

the `designs` alias works but `x:battery_amps` throws error

```
{
  "errors": [
    {
      "message": "Cannot return null for non-nullable field TeamDesign.battery_amps.",
      "locations": [
        {
          "line": 4,
          "column": 5
        }
      ],
      "path": [
        "designs",
        0,
        "x"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "Error: Cannot return null for non-nullable field TeamDesign.battery_amps.",
            "    at completeValue (/app/node_modules/graphql/execution/execute.js:559:13)",
            "    at resolveField (/app/node_modules/graphql/execution/execute.js:472:19)",
            "    at executeFields (/app/node_modules/graphql/execution/execute.js:292:18)",
            "    at collectAndExecuteSubfields (/app/node_modules/graphql/execution/execute.js:748:10)",
            "    at completeObjectValue (/app/node_modules/graphql/execution/execute.js:738:10)",
            "    at completeValue (/app/node_modules/graphql/execution/execute.js:590:12)",
            "    at completeValue (/app/node_modules/graphql/execution/execute.js:556:21)",
            "    at /app/node_modules/graphql/execution/execute.js:620:25",
            "    at Array.map (<anonymous>)",
            "    at safeArrayFrom (/app/node_modules/graphql/jsutils/safeArrayFrom.js:36:23)"
          ]
        }
      }
    }
  ],
  "data": null
}
```

**Comment by cramatt** [comment]
Tested with 1.2.3 and the above works find without an error.

**Comment by darrellwarde** [comment]
Hey @cramatt, this fix was actually addressed in 2.0.1, not 1.2.4. I realise it's a breaking change, but please upgrade to 2.0.1 to benefit from this fix.

**Comment by dmoree** [comment]
@darrellwarde Actually for `ID`, `Int`, and `Float` fields when composing object fields a resolver is being set that does not take advantage of the default field resolver even for `2.0.1`. I'm not sure what version @cramatt is using, but I've created #439 to address this.
